File: 346__Moving_Average_from_Data_Stream.py
class MovingAverage(object):

    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        self.data = []
        self.size = size


    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        if len(self.data) < self.size:
            self.data += [val]
            return sum(self.data) / float(len(self.data))
        else:
            self.data = self.data[1:]
            self.data += [val]
            return sum(self.data) / float(len(self.data))



# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)

# A Faster Solution
class MovingAverage(object):

    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        self.size = size
        self.vals = []
        self.count = 0.0
        self.ptr = 0

    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        self.vals.append(float(val))
        self.count += val

        if len(self.vals) > self.size:
            self.count -= self.vals[self.ptr]
            self.ptr += 1
            return self.count / self.size

        return self.count / len(self.vals)

File: Java/346__Moving_Average_from_Data_Stream.java
class MovingAverage {

    class ListNode {
        int val;
        ListNode next;
        public ListNode(int val) {
            this.val = val;
            next = null;
        }
    }

    ListNode dummy = null;
    ListNode cur = null;
    int total_size;
    int capacity;
    int total;

    /** Initialize your data structure here. */
    public MovingAverage(int size) {
        dummy = new ListNode(0);
        cur = dummy;
        total_size = 0;
        capacity = size;
        total = 0;
    }

    public double next(int val) {
        total_size++;
        int numToRemove = 0;
        if (total_size > capacity) {
            if (dummy.next == null) return 0.0;
            numToRemove = dummy.next.val;
            dummy.next = dummy.next.next;
            total_size--;
        }
        total += val;
        total -= numToRemove;
        cur.next = new ListNode(val);
        cur = cur.next;
        return (double) total / total_size;
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */
